File: frontend/src/types/taskTemplate.ts
export interface TaskTemplateExercise {
  id: string;
  templateId: string;
  exerciseId: string;
  exerciseName: string | null;
  orderIndex: number;
}

export interface TaskTemplate {
  id: string;
  ownerId: string;
  title: string;
  description: string | null;
  planId: string | null;
  metricName: string | null;
  metricUnit: string | null;
  metricTarget: number | null;
  requireComment: boolean;
  requirePhoto: boolean;
  requireVideo: boolean;
  requireSetsReps: boolean;
  requireDuration: boolean;
  requireMetricValue: boolean;
  requireDifficulty: boolean;
  requireWellbeing: boolean;
  exercises: TaskTemplateExercise[];
}

export interface TaskTemplateInput {
  title: string;
  description: string | null;
  plan_id: string | null;
  exercise_ids: string[];
  metric_name: string | null;
  metric_unit: string | null;
  metric_target: number | null;
  require_comment: boolean;
  require_photo: boolean;
  require_video: boolean;
  require_sets_reps: boolean;
  require_duration: boolean;
  require_metric_value: boolean;
  require_difficulty: boolean;
  require_wellbeing: boolean;
}

export interface TaskTemplateExerciseDto {
  id: string;
  template_id: string;
  exercise_id: string;
  exercise_name: string | null;
  order_index: number;
}

export interface TaskTemplateDto {
  id: string;
  owner_id: string;
  title: string;
  description: string | null;
  plan_id: string | null;
  metric_name: string | null;
  metric_unit: string | null;
  metric_target: number | null;
  require_comment: boolean;
  require_photo: boolean;
  require_video: boolean;
  require_sets_reps: boolean;
  require_duration: boolean;
  require_metric_value: boolean;
  require_difficulty: boolean;
  require_wellbeing: boolean;
  exercises: TaskTemplateExerciseDto[];
}

export function mapTaskTemplateExerciseDto(dto: TaskTemplateExerciseDto): TaskTemplateExercise {
  return {
    id: dto.id,
    templateId: dto.template_id,
    exerciseId: dto.exercise_id,
    exerciseName: dto.exercise_name,
    orderIndex: dto.order_index,
  };
}

export function mapTaskTemplateDto(dto: TaskTemplateDto): TaskTemplate {
  return {
    id: dto.id,
    ownerId: dto.owner_id,
    title: dto.title,
    description: dto.description,
    planId: dto.plan_id,
    metricName: dto.metric_name,
    metricUnit: dto.metric_unit,
    metricTarget: dto.metric_target,
    requireComment: dto.require_comment,
    requirePhoto: dto.require_photo,
    requireVideo: dto.require_video,
    requireSetsReps: dto.require_sets_reps,
    requireDuration: dto.require_duration,
    requireMetricValue: dto.require_metric_value,
    requireDifficulty: dto.require_difficulty,
    requireWellbeing: dto.require_wellbeing,
    exercises: dto.exercises.map(mapTaskTemplateExerciseDto),
  };
}

File: backend/app/schemas/task_template.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class TaskTemplateExerciseOut(BaseModel):
    id: UUID
    template_id: UUID
    exercise_id: UUID
    exercise_name: str | None
    order_index: int


class TaskTemplateIn(BaseModel):
    title: str = Field(min_length=1, max_length=150)
    description: str | None = Field(default=None, max_length=2000)
    plan_id: UUID | None = None
    exercise_ids: list[UUID] = Field(default_factory=list, max_length=30)
    metric_name: str | None = Field(default=None, max_length=100)
    metric_unit: str | None = Field(default=None, max_length=30)
    metric_target: float | None = None
    require_comment: bool = False
    require_photo: bool = False
    require_video: bool = False
    require_sets_reps: bool = False
    require_duration: bool = False
    require_metric_value: bool = False
    require_difficulty: bool = False
    require_wellbeing: bool = False


class TaskTemplateOut(BaseModel):
    id: UUID
    owner_id: UUID
    title: str
    description: str | None
    plan_id: UUID | None
    metric_name: str | None
    metric_unit: str | None
    metric_target: float | None
    require_comment: bool
    require_photo: bool
    require_video: bool
    require_sets_reps: bool
    require_duration: bool
    require_metric_value: bool
    require_difficulty: bool
    require_wellbeing: bool
    created_at: datetime
    updated_at: datetime
    exercises: list[TaskTemplateExerciseOut] = []
